metrics-runtime/src/registry/metric.rs

```diff
@@ -65,34 +66,63 @@ impl MetricRegistry {
             let scope = self.scope_registry.get(scope_handle);
-            let key = key.map_name(|name| scope.into_scoped(name));
 
-            let snapshot = value.snapshot();
-            named_values.push((key, snapshot));
+            match value.snapshot() {
+                ValueSnapshot::Single(measurement) => {
+                    let key = key.map_name(|name| scope.into_scoped(name));
+                    values.push((key, measurement));
+                }
+                ValueSnapshot::Multiple(mut measurements) => {
+                    for (subkey, measurement) in measurements.drain(..) {
+                        let scope = scope.clone();
+                        let subkey = subkey.map_name(|name| scope.into_scoped(name));
+                        values.push((subkey, measurement));
+                    }
+                }
+            }
         }
 
-        Snapshot::new(named_values)
+        Snapshot::new(values)
     }
 
     pub fn observe<O: Observer>(&self, observer: &mut O) {
         let metrics = self.metrics.load().deref().clone();
         for (id, value) in metrics.into_iter() {
             let (key, scope_handle, _) = id.into_parts();
             let scope = self.scope_registry.get(scope_handle);
-            let key = key.map_name(|name| scope.into_scoped(name));
 
-            match value.snapshot() {
-                ValueSnapshot::Counter(value) => observer.observe_counter(key, value),
-                ValueSnapshot::Gauge(value) => observer.observe_gauge(key, value),
-                ValueSnapshot::Histogram(stream) => stream.decompress_with(|values| {
+            let observe = |observer: &mut O, key, measurement| match measurement {
+                Measurement::Counter(value) => observer.observe_counter(key, value),
+                Measurement::Gauge(value) => observer.observe_gauge(key, value),
+                Measurement::Histogram(stream) => stream.decompress_with(|values| {
                     observer.observe_histogram(key.clone(), values);
                 }),
+            };
+
+            match value.snapshot() {
+                ValueSnapshot::Single(measurement) => {
+                    let key = key.map_name(|name| scope.into_scoped(name));
+                    observe(observer, key, measurement);
+                }
+                ValueSnapshot::Multiple(mut measurements) => {
+                    // Tack on the key name that this proxy was registered with to the scope so
+                    // that we can clone _that_, and then scope our individual measurements.
+                    let mut scope = scope.clone();
+                    let key_parts = key.into_parts();
+                    scope.add_part(key_parts.0);
+
+                    for (subkey, measurement) in measurements.drain(..) {
+                        let scope = scope.clone();
+                        let subkey = subkey.map_name(|name| scope.into_scoped(name));
```

Comment:
We should probably accept labels for proxy creation and then apply them here to the subkey after splitting them from the proxy key.

Reply:
This is also now done.